### ISSUE TITLE
Phase 24.4 — SocialPredictor: Trust Networks, Coalition Dynamics & Group Behavior

### ISSUE THREAD

**Issue by web3guru888** [web3guru888/asi-build#549]
### Overview

The `SocialPredictor` models and predicts **social dynamics** across multi-agent systems: trust evolution, coalition formation and dissolution, reputation propagation, and emergent group behaviors. It combines **game-theoretic analysis** (Shapley values for fair coalition allocation), **trust network propagation** (Marsh 1994), and **evolutionary game dynamics** (replicator equations) to forecast how social structures will evolve.

This component transforms point-in-time social snapshots into predictive models—answering "who will ally with whom?", "whose trust is eroding?", and "what group behaviors will emerge?"

---

### Data Structures

```python
class RelationshipType(enum.Enum):
    """Types of social relationships."""
    COOPERATIVE = "cooperative"
    COMPETITIVE = "competitive"
    NEUTRAL = "neutral"
    ADVERSARIAL = "adversarial"
    DEPENDENT = "dependent"

class TrustLevel(enum.Enum):
    """Discrete trust levels (Marsh 1994 adaptation)."""
    BLIND_TRUST = "blind_trust"      # > 0.9
    HIGH_TRUST = "high_trust"        # 0.7 - 0.9
    MODERATE_TRUST = "moderate"      # 0.4 - 0.7
    LOW_TRUST = "low_trust"          # 0.1 - 0.4
    DISTRUST = "distrust"            # < 0.1

@dataclass(frozen=True)
class AgentRelationship:
    """Directed relationship between two agents."""
    source_id: str
    target_id: str
    relationship_type: RelationshipType
    trust_score: float           # [0, 1] — directed trust
    interaction_count: int
    last_interaction_ms: int
    reciprocity: float           # Correlation of mutual cooperation [-1, 1]

@dataclass(frozen=True)
class Coalition:
    """A coalition of cooperating agents."""
    coalition_id: str
    members: FrozenSet[str]
    formation_ms: int
    stability: float             # [0, 1] — probability of persisting
    shared_goals: tuple[str, ...]
    shapley_values: Mapping[str, float]  # Member → marginal contribution
    total_value: float

@dataclass(frozen=True)
class SocialGraph:
    """Complete social structure snapshot."""
    agents: tuple[str, ...]
    relationships: tuple[AgentRelationship, ...]
    trust_matrix: Mapping[str, Mapping[str, float]]  # agent → agent → trust
    coalitions: tuple[Coalition, ...]
    clustering_coefficient: float  # Graph clustering
    average_trust: float
    timestamp_ms: int

@dataclass(frozen=True)
class SocialPrediction:
    """A prediction about future social dynamics."""
    prediction_type: str         # "coalition_formation" | "trust_change" | "defection"
    agents_involved: tuple[str, ...]
    probability: float
    time_horizon_ms: int
    reasoning: str               # Natural language explanation
    confidence: float

@dataclass(frozen=True)
class SocialPredictorConfig:
    """Configuration for social prediction."""
    trust_decay_rate: float = 0.01       # Trust decay per time unit without interaction
    trust_propagation_depth: int = 3     # Hops for transitive trust
    trust_propagation_discount: float = 0.5  # Decay per hop
    shapley_sampling_n: int = 1000       # Monte Carlo samples for Shapley
    coalition_min_stability: float = 0.3  # Below this → predict dissolution
    replicator_dt: float = 0.1           # Evolutionary dynamics time step
```

---

### Protocol

```python
@runtime_checkable
class SocialPredictor(Protocol):
    """Predict social dynamics, coalitions, and trust evolution."""

    async def predict_behavior(
        self,
        agent_id: str,
        context: Mapping[str, Any],
    ) -> tuple[SocialPrediction, ...]:
        """Predict agent's social behavior in context.

        Combines trust state, coalition membership, intention model,
        and evolutionary fitness to forecast actions.
        """
        ...

    async def forecast_coalition(
        self,
        agents: tuple[str, ...],
        time_horizon_ms: int,
    ) -> tuple[Coalition, ...]:
        """Forecast coalition structures over time horizon.

        Algorithm:
        1. Compute pairwise trust × goal alignment scores
        2. Find stable coalition structures (core of cooperative game)
        3. Compute Shapley values for each coalition
        4. Filter by stability threshold
        """
        ...

    async def update_trust(
        self,
        source_id: str,
        target_id: str,
        outcome: float,          # [-1, 1] — negative=betrayal, positive=cooperation
    ) -> float:
        """Update directed trust after interaction outcome.

        trust_new = trust_old + α × (outcome - trust_old)
        where α = learning_rate, modulated by surprise.
        """
        ...

    async def propagate_trust(self) -> int:
        """Propagate trust transitively through the social graph.

        Transitive trust: T(A,C) = T(A,B) × T(B,C) × discount^hop
        Returns number of trust edges updated.
        """
        ...

    async def get_social_graph(self) -> SocialGraph:
        """Get current social graph snapshot."""
        ...

    async def get_social_dynamics(
        self,
        time_steps: int = 10,
    ) -> tuple[SocialGraph, ...]:
        """Simulate social dynamics forward using replicator equations.

        dx_i/dt = x_i × (f_i - φ)
        where f_i = fitness of strategy i, φ = average fitness
        """
        ...

    async def compute_shapley_values(
        self,
        agents: tuple[str, ...],
        value_fn: Any,
    ) -> Mapping[str, float]:
        """Compute Shapley values for coalition allocation.

        φ_i = Σ_{S ⊆ N\\{i}} [|S|!(|N|-|S|-1)! / |N|!] × [v(S∪{i}) - v(S)]
        Approximated via Monte Carlo sampling.
        """
        ...
```

---

### Implementation: `GraphSocialPredictor`

```python
class GraphSocialPredictor:
    """Graph-based social dynamics prediction.

    Internal state:
        _trust_matrix: dict[str, dict[str, float]]   # Directed trust
        _relationships: dict[tuple[str,str], AgentRelationship]
        _coalitions: list[Coalition]
        _interaction_history: dict[tuple[str,str], deque[float]]
        _perspective_taker: PerspectiveTaker   # 24.3
        _consensus_voting: ConsensusVoting     # 12.4
        _config: SocialPredictorConfig
        _lock: asyncio.Lock

    Trust update algorithm:
        1. Compute surprise = |outcome - expected_trust|
        2. Adaptive learning rate: α = base_α × (1 + surprise)
        3. Update: trust_new = clamp(trust_old + α × (outcome - trust_old), 0, 1)
        4. Update relationship type based on trust thresholds

    Trust propagation (BFS):
        1. For each agent pair (A, C) with no direct interaction:
        2. Find all paths A → B₁ → ... → Bₙ → C of length ≤ depth
        3. Transitive trust = max over paths of Π(trust_edge × discount^hop)
        4. Update trust_matrix[A][C] = max(direct, transitive)

    Coalition formation (cooperative game theory):
        1. Compute characteristic function v(S) for each subset S
        2. Find core: allocations where no sub-coalition can improve
        3. If core is empty, use Shapley value as fair allocation
        4. Stability = min(excess) over all blocking coalitions

    Replicator dynamics:
        Uses discrete-time replicator equation for strategy evolution:
        x_i(t+1) = x_i(t) × f_i(t) / φ(t)
        where f_i = expected payoff of strategy i, φ = average payoff
    """
```

---

### Integration Points

| Component | Direction | Interface |
|-----------|-----------|-----------|
| PerspectiveTaker 24.3 | ← input | Perspective informs trust assessment |
| ConsensusVoting 12.4 | ← input | Voting patterns indicate coalitions |
| AffectiveOrchestrator 21.5 | ← input | Emotional state modulates trust |
| IntentionRecognizer 24.2 | ← input | Intentions predict cooperation/defection |
| SocialOrchestrator 24.5 | → output | SocialGraph + predictions |
| NegotiationEngine 12.2 | → output | Trust scores inform negotiation strategy |

---

### Prometheus Metrics

| Metric | Type | Description |
|--------|------|-------------|
| `asi_social_trust_updates_total` | Counter | Trust score updates |
| `asi_social_coalitions_active` | Gauge | Number of active coalitions |
| `asi_social_trust_propagation_seconds` | Histogram | Trust propagation latency |
| `asi_social_shapley_computation_seconds` | Histogram | Shapley value computation time |
| `asi_social_prediction_accuracy` | Gauge | Rolling prediction accuracy |

---

### Test Targets (12)

| # | Test | What it verifies |
|---|------|-----------------|
| 1 | `test_trust_update_learning_rate` | Trust moves toward outcome at rate α |
| 2 | `test_trust_update_clamp_bounds` | Trust stays in [0, 1] |
| 3 | `test_trust_propagation_transitive` | T(A,C) computed from T(A,B)×T(B,C) |
| 4 | `test_trust_propagation_depth_limit` | No propagation beyond configured depth |
| 5 | `test_shapley_values_efficiency` | Shapley values sum to v(N) |
| 6 | `test_shapley_values_symmetry` | Equal contributors get equal share |
| 7 | `test_coalition_stability_core` | Stable coalition satisfies core constraints |
| 8 | `test_coalition_dissolution_prediction` | Low-stability coalition → dissolution predicted |
| 9 | `test_replicator_dynamics_convergence` | Evolutionary dynamics converge to ESS |
| 10 | `test_social_graph_snapshot_consistency` | Snapshot reflects current trust/coalitions |
| 11 | `test_trust_decay_without_interaction` | Trust decays toward 0.5 over time |
| 12 | `test_null_social_predictor_no_ops` | NullSocialPredictor returns empty graph |

---

### Implementation Order

1. `RelationshipType` + `TrustLevel` enums
2. `AgentRelationship` + `Coalition` frozen dataclasses
3. `SocialGraph` + `SocialPrediction` frozen dataclasses
4. `SocialPredictorConfig` frozen dataclass
5. `SocialPredictor` Protocol
6. `GraphSocialPredictor.__init__` (trust matrix, relationships, lock)
7. `update_trust` — adaptive learning rate update
8. `propagate_trust` — BFS transitive propagation
9. `compute_shapley_values` — Monte Carlo approximation
10. `forecast_coalition` — core computation + stability
11. `predict_behavior` — composite social prediction
12. `get_social_dynamics` — replicator equation simulation
13. `NullSocialPredictor` no-op
14. `make_social_predictor()` factory + metrics + tests

**Comment by web3guru888** [comment]
## Implementation Notes — SocialPredictor

### Trust Update with Adaptive Learning Rate

```python
async def update_trust(self, source_id, target_id, outcome):
    current = self._trust_matrix.get(source_id, {}).get(target_id, 0.5)
    surprise = abs(outcome - current)
    alpha = self._config.base_learning_rate * (1.0 + surprise)
    new_trust = current + alpha * (outcome - current)
    new_trust = max(0.0, min(1.0, new_trust))  # Clamp
    self._trust_matrix.setdefault(source_id, {})[target_id] = new_trust
    return new_trust
```

**Key design**: Surprise-modulated learning rate means unexpected betrayals (outcome=-1 when trust=0.9) cause large trust drops, while expected cooperation causes small updates. This matches empirical findings on trust dynamics.

### Trust Propagation — BFS Implementation

```python
async def propagate_trust(self):
    updated = 0
    for source in self._trust_matrix:
        visited = {source}
        queue = deque([(source, 1.0, 0)])  # (node, accumulated_trust, depth)
        while queue:
            current, acc_trust, depth = queue.popleft()
            if depth >= self._config.trust_propagation_depth:
                continue
            for target, direct_trust in self._trust_matrix.get(current, {}).items():
                if target in visited or target == source:
                    continue
                visited.add(target)
                transitive = acc_trust * direct_trust * self._config.trust_propagation_discount
                existing = self._trust_matrix.get(source, {}).get(target, 0.0)
                if transitive > existing:
                    self._trust_matrix.setdefault(source, {})[target] = transitive
                    updated += 1
                queue.append((target, transitive, depth + 1))
    return updated
```

### Monte Carlo Shapley — Convergence Check

```python
async def compute_shapley_values(self, agents, value_fn):
    n = len(agents)
    shapley = {a: 0.0 for a in agents}
    for _ in range(self._config.shapley_sampling_n):
        perm = random.sample(agents, n)
        prev_value = 0.0
        for i, agent in enumerate(perm):
            coalition = frozenset(perm[:i+1])
            curr_value = await value_fn(coalition)
            shapley[agent] += (curr_value - prev_value) / self._config.shapley_sampling_n
            prev_value = curr_value
    return shapley
```

**Efficiency property**: `sum(shapley.values()) ≈ value_fn(frozenset(agents))` — verify in tests.

### Coalition Core Stability

```python
def _check_core_stability(self, coalition, shapley_values, value_fn):
    min_excess = float('inf')
    members = list(coalition.members)
    for size in range(1, len(members)):
        for sub in itertools.combinations(members, size):
            sub_set = frozenset(sub)
            sub_value = value_fn(sub_set)  # What sub-coalition could get alone
            allocated = sum(shapley_values.get(a, 0) for a in sub)
            excess = allocated - sub_value  # Positive = stable, negative = blocking
            min_excess = min(min_excess, excess)
    return max(0.0, min_excess / coalition.total_value) if coalition.total_value > 0 else 0.0
```

### Replicator Dynamics Discrete-Time

```python
def _replicator_step(self, strategy_proportions, payoff_matrix):
    fitness = payoff_matrix @ strategy_proportions  # Expected payoff per strategy
    avg_fitness = strategy_proportions @ fitness
    new_proportions = strategy_proportions * fitness / avg_fitness
    new_proportions = np.clip(new_proportions, 1e-10, None)  # Prevent extinction
    return new_proportions / new_proportions.sum()  # Renormalize
```

### mypy Narrowing

| Expression | Narrowed Type | Note |
|-----------|--------------|------|
| `self._trust_matrix.get(s, {}).get(t)` | `float \| None` | Default to 0.5 on None |
| `RelationshipType.COOPERATIVE` | Literal | Match/case exhaustive |
| `coalition.shapley_values` | `Mapping[str, float]` | Frozen — read-only |

### Test Skeletons

```python
async def test_shapley_values_efficiency():
    predictor = GraphSocialPredictor(config=SocialPredictorConfig())
    agents = ("A", "B", "C")
    async def value_fn(coalition):
        return len(coalition) * 10.0  # Linear value function
    shapley = await predictor.compute_shapley_values(agents, value_fn)
    assert abs(sum(shapley.values()) - 30.0) < 1.0  # Efficiency: sums to v(N)

async def test_trust_propagation_transitive():
    predictor = GraphSocialPredictor(config=SocialPredictorConfig(trust_propagation_depth=2))
    await predictor.update_trust("A", "B", 0.8)
    await predictor.update_trust("B", "C", 0.9)
    await predictor.propagate_trust()
    graph = await predictor.get_social_graph()
    # T(A,C) should be approximately 0.8 * 0.9 * 0.5 = 0.36
    assert 0.3 < graph.trust_matrix["A"]["C"] < 0.4
```
